src/set_theory/zfc/basic.lean

```diff
@@ -857,6 +866,17 @@ theorem map_unique {f : Set.{u} → Set.{u}} [H : definable 1 f] {x z : Set.{u}}
 λ h, ⟨λ y yx, let ⟨z, zx, ze⟩ := mem_image.1 yx in ze ▸ pair_mem_prod.2 ⟨zx, h z zx⟩,
      λ z, map_unique⟩⟩
 
+/-- Given a predicate `p` on ZFC sets. `hereditarily p x` means that `x` has property `p` and the
+members of `x` are all `hereditarily p`. -/
+def hereditarily (p : Set → Prop) (x : Set) : Prop :=
+x.rec_on (λ x h, p x ∧ ∀ y ∈ x, h y H)
+
+lemma hereditarily_iff {p : Set → Prop} {x : Set} :
```

Comment:
`hereditary p x → p x` and `hereditary p x → y ∈ x → hereditary p y` would be very nice as dot notation lemmas.